youtube_dl/extractor/animelab.py

```diff
@@ -0,0 +1,231 @@
+
+                if quality:
+                    height = int_or_none(self._search_regex(r'(\d+)p?$', quality, 'Video format height', default=None, fatal=False))
+                else:
+                    height = None
+
+                if height is None:
+                    self.report_warning('Could not get height of video')
+                else:
+                    current_format['height'] = height
+
+                format_id_parts = []
+
+                format_id_parts.append(str_or_none(video_instance.get('id')))
+
+                if is_hardsubbed is not None:
+                    if is_hardsubbed:
+                        format_id_parts.append('yeshardsubbed')
+                    else:
+                        format_id_parts.append('nothardsubbed')
+
+                format_id_parts.append(current_format['language'])
+
+                format_id = '_'.join([x for x in format_id_parts if x is not None])
+                current_format['format_id'] = format_id
+
+                formats.append(current_format)
+
+        # due to Python's sorting algorithm having the property of being 'stable', the following
+        # means that videos are sorted by height, but in the case of the same height,
+        # are sorted by language
+        formats.sort(key=lambda k: str.lower(k.get('language') or ''))
+        formats.sort(key=lambda k: k.get('height') or 0)
+
+        return formats
+
+    def _extract_video_from_collection(self, webpage, video_collection, position):
+        raw_data = video_collection[position]['videoEntry']
+
+        video_id = str_or_none(raw_data['id'])
+
+        # create a title from many sources (while grabbing other info)
+        # TODO use more fallback sources to get some of these
+        series = raw_data.get('showTitle')
+        video_type = raw_data.get('videoEntryType', {}).get('name')
+        episode_number = raw_data.get('episodeNumber')
+        episode_name = raw_data.get('name')
+
+        title_parts = (series, video_type, episode_number, episode_name)
+        title = self._create_title(webpage, title_parts)
+
+        description = raw_data.get('synopsis') or self._og_search_description(webpage, default=None)
+
+        formats = self._get_available_formats(webpage, raw_data)
+
+        # TODO extract thumbnails and other things youtube-dl optionally wants
+
+        return {
+            'id': video_id,
+            'title': title,
+            'description': description,
+            'series': series,
+            'episode': episode_name,
+            'episode_number': int_or_none(episode_number),
+            'formats': formats,
+        }
+
+    def _extract_video_from_player(self, webpage):
+        video_collection = self._extract_video_collection_from_player(webpage)
+        position = self._extract_position_from_player(webpage)
+        return self._extract_video_from_collection(webpage, video_collection, position)
+
+
+class AnimeLabIE(AnimeLabBaseIE):
```

Comment:
You're right, in this case the base class is quite pointless, but if I want to add playlist support in the future, a base class just for login will be useful (so I will make the base class smaller rather than completely removing it).